### PR TITLE
Descriptors support

### PR DESCRIPTION
**Pull request by Cierpliwy**
Add descriptor support for Android platform

* Update dev dependencies of the project
* Add missing descriptor functionalities on JS and TypeScript side.
* Add Android implementation of the descriptors.
* Refactor parts of Android source code.
* Update documentation.

### CODE REVIEW

File: android/src/main/java/com/polidea/reactnativeble/BleModule.java [hunk 1]
```diff
@@ -1429,17 +1850,24 @@ private void cleanServicesAndCharacteristicsForDevice(@NonNull Device device) {
...
+            if (service.getDeviceId().equals(device.getDeviceId())) {
                 discoveredServices.remove(key);
             }
         }
         for (int i = discoveredCharacteristics.size() - 1; i >=0; i--) {
             int key = discoveredCharacteristics.keyAt(i);
             Characteristic characteristic = discoveredCharacteristics.get(key);
 
-            if (characteristic.getService().getDevice().getNativeDevice().getMacAddress().equals(device.getNativeDevice().getMacAddress())) {
+            if (characteristic.getDeviceId().equals(device.getDeviceId())) {
```
Comment (dariuszseweryn) [inline android/src/main/java/com/polidea/reactnativeble/BleModule.java]: Some of the changes seem not related to the topic of the PR. Just sayin'

File: android/src/main/java/com/polidea/reactnativeble/errors/BleErrorCode.java [hunk 2]
```diff
@@ -44,6 +44,7 @@
     DescriptorNotFound(503),
     DescriptorsNotDiscovered(504),
     DescriptorInvalidDataFormat(505),
+    DescriptorWriteNotAllowed(506),
```
[the remark this answers is not in the document]
Comment (Cierpliwy) [reply android/src/main/java/com/polidea/reactnativeble/errors/BleErrorCode.java]: As far as I know, all descriptors are readable. Inside Characteristic Extended Properties descriptor, there are additional properties regarding descriptors, which are: “Reliable Write” and “Writable Auxiliaries”. Those are applicable based on UUID (basically description provided by the specification). I didn’t find anything related to read permissions on descriptors in the specification.

Regardless, the above error is more for and OS permissions than Bluetooth’s permissions.

File: android/src/main/java/com/polidea/reactnativeble/wrapper/Descriptor.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import com.polidea.reactnativeble.utils.ByteUtils;
+import com.polidea.reactnativeble.utils.IdGenerator;
+import com.polidea.reactnativeble.utils.IdGeneratorKey;
+import com.polidea.reactnativeble.utils.UUIDConverter;
+import com.polidea.rxandroidble.internal.RxBleLog;
+
+public class Descriptor {
+
+    private interface Metadata {
+        String ID = "id";
```
[the remark this answers is not in the document]
Comment (Cierpliwy) [reply android/src/main/java/com/polidea/reactnativeble/wrapper/Descriptor.java]: `final` is redundant in interface definition :)

File: android/src/main/java/com/polidea/reactnativeble/BleModule.java (hunk 1 above)
Comment (Cierpliwy) [reply android/src/main/java/com/polidea/reactnativeble/BleModule.java]: I agree I need to pick a different strategy for pull requests. What I want to achieve is consistent `master` branch at all times. Probably next time I will just send smaller PR to a different branch and merge it to `master` afterward.
